## Imports

In [1]:
from google.colab import drive
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

# Google Drive Connection
For this project, I will store the data on my personal google drive and below set up a connection to it. So, I can load the CSV files as they are from a system folder.

In [2]:
drive.mount('/gdrive')

Mounted at /gdrive


In [3]:
data_folder =  '/gdrive/My Drive/Data Science Career Track/Unit 7/capstone_two_project_folder/data/'

# Data Colleciton


## Data Loading

In [4]:
file_assessments  = data_folder + 'assessments.csv'
file_courses = data_folder + 'courses.csv'
file_student_assessment = data_folder + 'studentAssessment.csv'
file_student_info = data_folder + 'studentInfo.csv'
file_student_register = data_folder + 'studentRegistration.csv'
file_student_vle = data_folder + 'studentVle.csv'
file_vle = data_folder + 'vle.csv'

In [5]:
raw_assessments = pd.read_csv(file_assessments)
raw_courses = pd.read_csv(file_courses)
raw_student_assessment = pd.read_csv(file_student_assessment)
raw_student_info = pd.read_csv(file_student_info)
raw_student_register = pd.read_csv(file_student_register)
raw_student_vle = pd.read_csv(file_student_vle)
raw_vle = pd.read_csv(file_vle)

In [6]:
raw_assessments.head()

code_module code_presentation  id_assessment assessment_type   date  weight
0         AAA             2013J           1752             TMA   19.0    10.0
1         AAA             2013J           1753             TMA   54.0    20.0
2         AAA             2013J           1754             TMA  117.0    20.0
3         AAA             2013J           1755             TMA  166.0    20.0
4         AAA             2013J           1756             TMA  215.0    30.0

In [7]:
 raw_assessments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code_module        206 non-null    object 
 1   code_presentation  206 non-null    object 
 2   id_assessment      206 non-null    int64  
 3   assessment_type    206 non-null    object 
 4   date               195 non-null    float64
 5   weight             206 non-null    float64
dtypes: float64(2), int64(1), object(3)
memory usage: 9.8+ KB


In [8]:
raw_courses.head()

code_module code_presentation  module_presentation_length
0         AAA             2013J                         268
1         AAA             2014J                         269
2         BBB             2013J                         268
3         BBB             2014J                         262
4         BBB             2013B                         240

In [9]:
raw_courses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 3 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   code_module                 22 non-null     object
 1   code_presentation           22 non-null     object
 2   module_presentation_length  22 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 656.0+ bytes


In [10]:
raw_student_assessment.head()

id_assessment  id_student  date_submitted  is_banked  score
0           1752       11391              18          0   78.0
1           1752       28400              22          0   70.0
2           1752       31604              17          0   72.0
3           1752       32885              26          0   69.0
4           1752       38053              19          0   79.0

In [11]:
raw_student_assessment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173912 entries, 0 to 173911
Data columns (total 5 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id_assessment   173912 non-null  int64  
 1   id_student      173912 non-null  int64  
 2   date_submitted  173912 non-null  int64  
 3   is_banked       173912 non-null  int64  
 4   score           173739 non-null  float64
dtypes: float64(1), int64(4)
memory usage: 6.6 MB


In [12]:
raw_student_info.head()

code_module code_presentation  ...  disability final_result
0         AAA             2013J  ...           N         Pass
1         AAA             2013J  ...           N         Pass
2         AAA             2013J  ...           Y    Withdrawn
3         AAA             2013J  ...           N         Pass
4         AAA             2013J  ...           N         Pass

[5 rows x 12 columns]

In [13]:
raw_student_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   code_module           32593 non-null  object
 1   code_presentation     32593 non-null  object
 2   id_student            32593 non-null  int64 
 3   gender                32593 non-null  object
 4   region                32593 non-null  object
 5   highest_education     32593 non-null  object
 6   imd_band              31482 non-null  object
 7   age_band              32593 non-null  object
 8   num_of_prev_attempts  32593 non-null  int64 
 9   studied_credits       32593 non-null  int64 
 10  disability            32593 non-null  object
 11  final_result          32593 non-null  object
dtypes: int64(3), object(9)
memory usage: 3.0+ MB


In [14]:
raw_student_register.head()

code_module code_presentation  ...  date_registration  date_unregistration
0         AAA             2013J  ...             -159.0                  NaN
1         AAA             2013J  ...              -53.0                  NaN
2         AAA             2013J  ...              -92.0                 12.0
3         AAA             2013J  ...              -52.0                  NaN
4         AAA             2013J  ...             -176.0                  NaN

[5 rows x 5 columns]

In [15]:
raw_student_register.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   code_module          32593 non-null  object 
 1   code_presentation    32593 non-null  object 
 2   id_student           32593 non-null  int64  
 3   date_registration    32548 non-null  float64
 4   date_unregistration  10072 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 1.2+ MB


In [16]:
raw_student_vle.head()

code_module code_presentation  id_student  id_site  date  sum_click
0         AAA             2013J       28400   546652   -10          4
1         AAA             2013J       28400   546652   -10          1
2         AAA             2013J       28400   546652   -10          1
3         AAA             2013J       28400   546614   -10         11
4         AAA             2013J       28400   546714   -10          1

In [17]:
raw_student_vle.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10655280 entries, 0 to 10655279
Data columns (total 6 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   code_module        object
 1   code_presentation  object
 2   id_student         int64 
 3   id_site            int64 
 4   date               int64 
 5   sum_click          int64 
dtypes: int64(4), object(2)
memory usage: 487.8+ MB


There is this one issue here. On the documentation, it says that `"sum_click – the number of times a student interacts with the material in that day"`. However, the first three rows are identical to the VLE material and the student but have different `sum_click` values.


In [18]:
raw_vle.head()

id_site code_module code_presentation activity_type  week_from  week_to
0   546943         AAA             2013J      resource        NaN      NaN
1   546712         AAA             2013J     oucontent        NaN      NaN
2   546998         AAA             2013J      resource        NaN      NaN
3   546888         AAA             2013J           url        NaN      NaN
4   547035         AAA             2013J      resource        NaN      NaN

In [19]:
raw_vle.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6364 entries, 0 to 6363
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id_site            6364 non-null   int64  
 1   code_module        6364 non-null   object 
 2   code_presentation  6364 non-null   object 
 3   activity_type      6364 non-null   object 
 4   week_from          1121 non-null   float64
 5   week_to            1121 non-null   float64
dtypes: float64(2), int64(1), object(3)
memory usage: 298.4+ KB


## Data Explore

### `student_vle`

First, I will have a look at the issue about `sum_click` and try to understand the issue and group the data by columns `'code_module', 'code_presentation', 'id_student', 'id_site', 'date'`

In [20]:
grouped_student_vle = raw_student_vle.groupby(
    by=['code_module', 'code_presentation', 'id_student', 'id_site', 'date']
    ).sum('sum_click').reset_index()

In [21]:
num_rows_raw = raw_student_vle.shape[0] 
num_rows_grouped = grouped_student_vle.shape[0]
num_rows_diff = num_rows_raw - num_rows_grouped
(num_rows_raw, num_rows_grouped, num_rows_diff, (num_rows_diff) * 100 / num_rows_raw)

(10655280, 8459320, 2195960, 20.609125241194974)

The total num of rows in the raw table is 10'655'280  
If the data is groupped by all the columns except `sum_clicks` this is reduced to 8'459'320  
This means a **%20.6** shrinkage and the difference is 2'195'960  

---  

At this point I will assume that the actual total num of clicks is the groupped one. (Actually got reply from Martin Hols -one of data providers- and he suggested the same way))

In [22]:
# Duplicate entries
student_vle_duplicate = raw_student_vle.merge(
    grouped_student_vle,
    indicator = True,
    how='left',
    ).loc[lambda x : x['_merge']!='both']

Cross-check of the group by function, if the summation is correct.  
I will randomly select some course (means `code_module`, `code_presentation`),  student, site and date from `student_vle_duplicate`, then find the entry in both raw and grouped tables and compare. So I expect to find more than one row for each student on the `raw_student_vle` table and sum of the `sum_clicks` will be equal to the `sum_clicks` on the `grouped_student_vle`

In [23]:
rand_idx = np.random.choice(student_vle_duplicate.index)
rand_row = student_vle_duplicate.loc[rand_idx]

multiple = raw_student_vle[
                (raw_student_vle['code_module'] == rand_row['code_module']) &
                (raw_student_vle['code_presentation'] == rand_row['code_presentation']) &
                (raw_student_vle['id_student'] == rand_row['id_student']) &
                (raw_student_vle['id_site'] == rand_row['id_site']) &
                (raw_student_vle['date'] == rand_row['date']) 
               ]

single = grouped_student_vle[
                (grouped_student_vle['code_module'] == rand_row['code_module']) &
                (grouped_student_vle['code_presentation'] == rand_row['code_presentation']) &
                (grouped_student_vle['id_student'] == rand_row['id_student']) &
                (grouped_student_vle['id_site'] == rand_row['id_site']) &
                (grouped_student_vle['date'] == rand_row['date']) 
               ]

multiple, single

(        code_module code_presentation  id_student  id_site  date  sum_click
 2358222         CCC             2014B      585393   729680   206          1
 2358225         CCC             2014B      585393   729680   206          1,
         code_module code_presentation  id_student  id_site  date  sum_click
 1608907         CCC             2014B      585393   729680   206          2)

Sum process looks successful and from now on `grouped_student_vle` will be used instead of `raw_student_vle`

### `student_info`

How to handle the `num_of_prev_attemts` column?   

In [24]:
raw_student_info['num_of_prev_attempts'].unique()

array([0, 1, 2, 4, 3, 5, 6])

So there are some students with up to 6th trial.
    
Does their `id_student`change by trial?


In [25]:
raw_student_info['id_student'].unique().shape[0], raw_student_info.shape[0]

(28785, 32593)

It doesn't. So how should these former attempts be handled? As an additional effect on final trial's result or as an individual trial without any connection?  

First o all, data is prepared for the years 2013 and 2014. Are there any students who is taking any course second (or more) time in the first semester of 2013 

In [26]:
raw_student_info[
                 (raw_student_info['num_of_prev_attempts'] > 1) & 
                 (raw_student_info['code_presentation'] == '2013B')
                 ]

code_module code_presentation  ...  disability final_result
748           BBB             2013B  ...           N         Fail
750           BBB             2013B  ...           Y    Withdrawn
754           BBB             2013B  ...           N         Pass
764           BBB             2013B  ...           N         Fail
773           BBB             2013B  ...           Y         Fail
...           ...               ...  ...         ...          ...
23706         FFF             2013B  ...           N    Withdrawn
23760         FFF             2013B  ...           N    Withdrawn
23797         FFF             2013B  ...           N         Pass
23832         FFF             2013B  ...           N    Withdrawn
23868         FFF             2013B  ...           N         Pass

[185 rows x 12 columns]

There are such rows, so we don't have any information about their former attempt. Therefore, all attempts will be handled on their own.  
I also need to find out if there are any students who takes more than one course in a semester.  


In [27]:
students_with_multip_course = raw_student_info[raw_student_info.duplicated(subset=['id_student', 'code_presentation'], keep=False)].sort_values('id_student')
students_with_multip_course.head()

code_module code_presentation  ...  disability final_result
20418         EEE             2014B  ...           N         Pass
8659          CCC             2014B  ...           N         Pass
6372          BBB             2014J  ...           N         Pass
17569         DDD             2014J  ...           N    Withdrawn
16338         DDD             2014B  ...           N    Withdrawn

[5 rows x 12 columns]

In [28]:
len(students_with_multip_course['id_student'].unique())

1071

There are 1071 student who take more than one course at a semester.  

# Data Joining

Eventually, all the tables will be joined to `student_info`. But `assessment` and `vle` tables will be preprocessed and joined to `student_assessment` and `student_vle` tables respectively.



### Preparing `student_assesseent`
These tables will bring up information about assessments, like if it's tutor marked assignmet (`assessment_type == 'TMA'`) or the student's score on the assignment.

In [29]:
student_assessment = (
    raw_student_assessment.set_index('id_assessment')
    .join(raw_assessments.set_index('id_assessment'))
).reset_index()

At this point there are multiple entries for a `code_module`, `code_presentation`, `id_student`. So this table can not be added to `student_info`. I need to come up with some aggregation for student assessment.

In [30]:
student_assessment.head()

id_assessment  id_student  date_submitted  ...  assessment_type  date weight
0           1752       11391              18  ...              TMA  19.0   10.0
1           1752       28400              22  ...              TMA  19.0   10.0
2           1752       31604              17  ...              TMA  19.0   10.0
3           1752       32885              26  ...              TMA  19.0   10.0
4           1752       38053              19  ...              TMA  19.0   10.0

[5 rows x 10 columns]

In [31]:
student_assessment.query('id_student == 38053')

id_assessment  id_student  date_submitted  ...  assessment_type   date weight
4              1752       38053              19  ...              TMA   19.0   10.0
505            1753       38053              64  ...              TMA   54.0   20.0
846            1754       38053             117  ...              TMA  117.0   20.0
1036           1755       38053             166  ...              TMA  166.0   20.0
1461           1756       38053             215  ...              TMA  215.0   30.0

[5 rows x 10 columns]

In [32]:
student_assessment.query('id_student == 11391')

id_assessment  id_student  date_submitted  ...  assessment_type   date weight
0              1752       11391              18  ...              TMA   19.0   10.0
501            1753       11391              53  ...              TMA   54.0   20.0
842            1754       11391             115  ...              TMA  117.0   20.0
1032           1755       11391             164  ...              TMA  166.0   20.0
1457           1756       11391             212  ...              TMA  215.0   30.0

[5 rows x 10 columns]

Looks like not all courses have all types of assessments 

In [33]:
student_assessment.groupby(['code_module', 'code_presentation']).nunique()

id_assessment  id_student  ...  date  weight
code_module code_presentation                             ...              
AAA         2013J                          5         365  ...     5       3
            2014J                          5         340  ...     5       3
BBB         2013B                         11        1369  ...     7       3
            2013J                         11        1710  ...     7       3
            2014B                         11        1205  ...     7       3
            2014J                          5        1793  ...     5       4
CCC         2014B                          9        1415  ...     8       6
            2014J                          9        1998  ...     8       6
DDD         2013B                         14        1065  ...    14       9
            2013J                          7        1507  ...     7       5
            2014B                          7         944  ...     7       5
            2014J                          7        1428  ...     6       4
EEE         2013J                          4         827  ...     4       2
            2014B                          4         536  ...     4       2
            2014J                          4         935  ...     4       2
FFF         2013B                         12        1371  ...     6       3
            2013J                         12        1874  ...     6       3
            2014B                         12        1206  ...     6       3
            2014J                         12        1843  ...     6       3
GGG         2013J                          9         819  ...     4       1
            2014B                          9         689  ...     4       1
            2014J                          9         604  ...     4       1

[22 rows x 8 columns]

Yes, they do and this varies from 4 to 14. It doesn't make sense to add all assessments information to the related students. It would end up having around 6 columns with NaN values up to ~97% of the data.
  
So, I need to engineer the features and come up with new features anonymous to assessment number in a course

In [34]:
# avg_score: Student's avg score calculated with weights
# first_score: Student's first score
# mean_score: Average of the student's all scores (calculated without weights)
# last_score: Student's last score
# score_std: Standard deviation of the student's scores
# min_weight_score: Student's score on min weighted assignment
# max_weight_score: Student's score on max weighted assignment
# submit_delta: Total difference from student's submit days to the actual deadline
# submit_delta_std: Standard deviation of `submit_delta`
# num_banked: Total number of banked scores
# has_exam: `True` if course has an exam and `False` if not
# num_of_assmnt: Total number of assessments student takes at the course

#### NaN date values on `Exam` column
Some exams have no date. And for such rows its commented as _"If the information about the final exam date is missing, it is at the end of the last presentation week."_ So I need to fill with final date of the presentation to the NaN values.  
Does all values with date NaN are exams or are there any other missing info?

In [35]:
student_assessment[student_assessment['date'].isna()]['assessment_type'].unique()

array(['Exam'], dtype=object)

Since all the NaN values are only on exams these valuese can be filled with final date from course table

In [36]:
_idx = ['code_module', 'code_presentation']
end_dates = raw_courses.set_index(_idx)['module_presentation_length']
student_assessment.set_index(_idx, inplace=True)
student_assessment['date'].fillna(end_dates, inplace=True)
student_assessment.reset_index(inplace=True)

In [37]:
student_assessment['date'].isna().sum()

0

Filled all NaN values on `date` column with presentation end date


#### Calculating student specific features

##### TODO: Find a better naming


In [101]:
# Setting up the new DataFrame
student_assessment_reduced = pd.DataFrame()

# The new Dataframe will be created by grouping and aggregations of some 
# features in `student_assessment` table
_grouper = ['id_student', 'code_presentation', 'code_module']
grouped_assessment = student_assessment.groupby(_grouper)

# Calculating `avg_score`
_score = student_assessment.set_index(_grouper).score
_weight = student_assessment.set_index(_grouper).weight
# Total weight is not 100 for all courses
_total_weight = (student_assessment
                 .groupby(_grouper)
                 .sum()
                 .weight)
student_assessment_reduced['avg_score'] = ((_score *_weight / _total_weight)
                                            .groupby(_grouper)
                                            .sum())

# Calculating `first_score`
student_assessment_reduced['first_score'] = (grouped_assessment
                                             .first()
                                             .score)

# Calculating `mean_score`
student_assessment_reduced['mean_score'] = (grouped_assessment
                                           .mean()
                                           .score)

# Calculating `last_score`
student_assessment_reduced['last_score'] = (grouped_assessment
                                            .last()
                                            .score)

# Calculating `score_std`
student_assessment_reduced['score_std'] = (grouped_assessment
                                           .std()
                                           .score)

# Calculating `min_weight_score`
student_assessment_reduced['min_weight_score'] = (grouped_assessment
                                                  .min('weight')
                                                  .score)

# Calculating `max_weight_score`
student_assessment_reduced['max_weight_score'] = (grouped_assessment
                                                  .max('weight')
                                                  .score)

# `submit_delta`, `submit_delta_std` 
delta = (student_assessment.set_index(_grouper).date - student_assessment.set_index(_grouper).date_submitted)
student_assessment_reduced['submit_delta'] = delta.groupby(by=_grouper).sum()
student_assessment_reduced['submit_delta_std'] = delta.groupby(by=_grouper).std()

# `num_banked`
student_assessment_reduced['num_banked'] = (grouped_assessment
                                            .sum()
                                            .is_banked)

# `has_exam`
student_assessment_reduced['has_exam'] = (student_assessment.set_index(_grouper)['assessment_type'] == 'Exam').groupby(_grouper).sum()

# `num_of_assmnt`
student_assessment_reduced['num_of_assmnt'] = (grouped_assessment
                                               .nunique()
                                               .id_assessment)

#### Cross check for aggregated values

In [102]:
student_assessment_reduced.loc[49119]

avg_score  first_score  ...  has_exam  num_of_assmnt
code_presentation code_module                          ...                         
2014B             CCC          95.705000         98.0  ...         1              9
                  DDD          71.222222         69.0  ...         0              2

[2 rows x 12 columns]

In [103]:
student_assessment.set_index('id_student').loc[49119]

code_module code_presentation  ...   date  weight
id_student                                ...               
49119              CCC             2014B  ...   32.0     9.0
49119              CCC             2014B  ...  102.0    22.0
49119              CCC             2014B  ...  151.0    22.0
49119              CCC             2014B  ...  200.0    22.0
49119              CCC             2014B  ...   18.0     2.0
49119              CCC             2014B  ...   67.0     7.0
49119              CCC             2014B  ...  137.0     8.0
49119              CCC             2014B  ...  207.0     8.0
49119              CCC             2014B  ...  241.0   100.0
49119              DDD             2014B  ...   25.0    10.0
49119              DDD             2014B  ...   53.0    12.5

[11 rows x 9 columns]

### Preparing `student_vle`
Thse table will bring up student activity on virtual learning platform. Such as, `activity_type` or `sum_click`

In [ ]:
student_vle = (
    grouped_student_vle.set_index(['code_module', 'code_presentation', 'id_site'])
    .join(raw_vle.set_index(['code_module', 'code_presentation', 'id_site']))
).reset_index()

In [ ]:
student_vle.head()

code_module code_presentation  id_site  ...  activity_type  week_from  week_to
0         AAA             2013J   546614  ...       homepage        NaN      NaN
1         AAA             2013J   546614  ...       homepage        NaN      NaN
2         AAA             2013J   546614  ...       homepage        NaN      NaN
3         AAA             2013J   546614  ...       homepage        NaN      NaN
4         AAA             2013J   546614  ...       homepage        NaN      NaN

[5 rows x 9 columns]

At this point there are multiple entries for a `code_module`, `code_presentation`, `id_student`. So this table can not be added to `student_info`. I need to come up with some aggregation for student activity

In [ ]:
## TODO: Find out aggregate student activity on `student_vle` 

### Preparing `courses` 
This table will bring up `module_presentation_length` informaiton

In [ ]:
courses = raw_courses.set_index(['code_module', 'code_presentation'])

### Preparing `student_register`

In [ ]:
student_register = raw_student_register.set_index(['code_module', 'code_presentation', 'id_student'])

### Final Join 

In [ ]:
attempt = (
    raw_student_info
    .set_index(['code_module', 'code_presentation', 'id_student'])
    .join(courses)
    .join(student_register)

) 

In [ ]:
attempt.head()

# Data Definition
* Column names
* Data types
* Description of the columns
* Counts and percents unique values
* Ranges of values

# Data Cleaning
* NA or missing values
* Duplicates

 `student_info`

How a previous attempt should be considered?  Is it someting students build up their understanding in a long time? Or just another individual attempt?
 
Can student's former attempts'  can be added to their final trial. In other words, does student's learn something from failed courses and use it on their next trial?